File: blockchain/Transaction.py
import hashlib
import json
import logging

from blockchain import crypto
from blockchain.address import try_address_from_public_key
from blockchain.economics import MIN_FEE_PER_BYTE, MIN_GAS_PRICE, san_to_units
from utils import canonical

logger = logging.getLogger(__name__)

# Single canonical JSON representation used for signing, verification and
# block hashing. Every component must use these exact settings.


class Transaction:

    # Fields that are protocol metadata and are never part of a user signature.
    META_FIELDS = ("signature", "fee")

    # Chain binding: the payload must name the chain it is valid on, and the
    # node rejects any transaction whose id does not match its own.
    CHAIN_ID_FIELD = "chain_id"

    # Payload keys that trigger code execution (and therefore need gas).
    EXECUTION_FIELDS = ("bytecode", "contract_code")

    # System commands for the validator registry (staking).
    VALIDATOR_COMMANDS = ("deposit", "undelegate", "withdraw", "evidence")

    # On-chain parameter changes need a 2/3 validator approval set.
    GOVERNANCE_COMMANDS = ("set_param",)

    def __init__(self, data, fee_rate=None):
        self.payload = self._normalize_payload(data)
        # The client cannot pick its own fee; the protocol recomputes it.
        self.payload.pop("fee", None)

        if not self.chain_id_of(self.payload):
            raise ValueError(
                f"Transaction payload must include a non-empty '{self.CHAIN_ID_FIELD}'"
            )
        error = self.validate_gas_fields(self.payload)
        if error:
            raise ValueError(error)
        error = self.validate_validator_command(self.payload)
        if error:
            raise ValueError(error)
        error = self.validate_governance_command(self.payload)
        if error:
            raise ValueError(error)
        error = self.validate_contract_code(self.payload)
        if error:
            raise ValueError(error)

        rate = MIN_FEE_PER_BYTE if fee_rate is None else int(fee_rate)
        self.fee = self.expected_fee(self.payload, rate)
        self.payload["fee"] = self.fee
        self.data = canonical.dumps_bytes(self.payload)

    @staticmethod
    def validator_command_of(payload: dict):
        command = payload.get("validator")
        return command if isinstance(command, dict) else None

    @classmethod
    def validate_validator_command(cls, payload: dict) -> str | None:
        """Structural rules for staking/slashing system transactions."""
        command = payload.get("validator")
        if command is None:
            return None
        if not isinstance(command, dict):
            return "'validator' must be an object"

        name = command.get("command")
        if name not in cls.VALIDATOR_COMMANDS:
            return f"unknown validator command: {name!r}"
        if cls.has_execution(payload):
            return "validator commands must not contain execution payloads"
        if "value" in payload or "receiver" in payload:
            return "validator commands must not be transfers"

        if name == "deposit":
            amount = command.get("amount")
            if isinstance(amount, bool) or not isinstance(amount, int) or amount <= 0:
                return "validator deposit requires a positive integer 'amount'"

        if name == "evidence":
            for field in ("vote_a", "vote_b"):
                vote = command.get(field)
                if not isinstance(vote, dict):
                    return f"evidence requires a '{field}' vote object"
                if not vote.get("public_key") or not vote.get("signature"):
                    return f"evidence '{field}' must be a signed vote"
        return None

    @staticmethod
    def validate_contract_code(payload: dict) -> str | None:
        """Structural rules for deploy/run payloads (audit fix: a non-string
        contract id used to crash execution and poison the mempool)."""
        code = payload.get("contract_code")
        if code is None:
            return None
        if not isinstance(code, dict):
            return "'contract_code' must be an object"

        command = code.get("command")
        if command not in ("deploy", "run"):
            return f"unknown contract command: {command!r}"

        contract_id = code.get("contract_id")
        if not isinstance(contract_id, str) or not contract_id:
            return "contract requires a non-empty string 'contract_id'"

        if command == "deploy":
            if not isinstance(code.get("pena_code"), str) and not isinstance(
                code.get("bytecode"), list
            ):
                return "deploy requires 'pena_code' or 'bytecode'"
            return None

        if not isinstance(code.get("function_name"), str) or not code["function_name"]:
            return "run requires a non-empty 'function_name'"
        if not isinstance(code.get("params", []), list):
            return "'params' must be a list"
        return None

    @staticmethod
    def governance_command_of(payload: dict):
        command = payload.get("governance")
        return command if isinstance(command, dict) else None

    @classmethod
    def validate_governance_command(cls, payload: dict) -> str | None:
        """Structural rules for on-chain parameter changes."""
        command = payload.get("governance")
        if command is None:
            return None
        if not isinstance(command, dict):
            return "'governance' must be an object"

        name = command.get("command")
        if name not in cls.GOVERNANCE_COMMANDS:
            return f"unknown governance command: {name!r}"
        if cls.has_execution(payload):
            return "governance commands must not contain execution payloads"
        if "value" in payload or "receiver" in payload:
            return "governance commands must not be transfers"
        if payload.get("validator") is not None:
            return "governance and validator commands cannot be combined"

        if not isinstance(command.get("name"), str) or not command.get("name"):
            return "governance set_param requires a 'name'"
        value = command.get("value")
        if isinstance(value, bool) or not isinstance(value, int):
            return "governance set_param requires an integer 'value'"

        approvals = command.get("approvals")
        if not isinstance(approvals, list) or not approvals:
            return "governance set_param requires at least one approval"
        for approval in approvals:
            if not isinstance(approval, dict):
                return "governance approvals must be objects"
            if not approval.get("public_key") or not approval.get("signature"):
                return "governance approvals must be signed by validators"
        return None

    @staticmethod
    def chain_id_of(payload: dict):
        return payload.get(Transaction.CHAIN_ID_FIELD)

    @staticmethod
    def has_execution(payload: dict) -> bool:
        return any(field in payload for field in Transaction.EXECUTION_FIELDS)

    @staticmethod
    def _int_field(payload: dict, name: str) -> int:
        value = payload.get(name, 0)
        if isinstance(value, bool) or not isinstance(value, int) or value < 0:
            raise ValueError(f"'{name}' must be a non-negative integer")
        return value

    @staticmethod
    def gas_limit_of(payload: dict) -> int:
        return Transaction._int_field(payload, "gas_limit")

    @staticmethod
    def gas_price_of(payload: dict) -> int:
        return Transaction._int_field(payload, "gas_price")

    @classmethod
    def validate_gas_fields(cls, payload: dict) -> str | None:
        """Gas rules: execution pays for computation, plain transfers do not."""
        try:
            gas_limit = cls._int_field(payload, "gas_limit")
            gas_price = cls._int_field(payload, "gas_price")
        except ValueError as exc:
            return str(exc)

        if cls.has_execution(payload):
            if gas_limit < 1:
                return "execution transactions require 'gas_limit' >= 1"
            if gas_price < MIN_GAS_PRICE:
                return f"'gas_price' must be at least {MIN_GAS_PRICE}"
        else:
            if gas_limit != 0 or gas_price != 0:
                return "transactions without execution must not set gas fields"
        return None

    @staticmethod
    def _normalize_payload(data) -> dict:
        """Accept a dict, a JSON string or JSON bytes and return a dict."""
        if isinstance(data, dict):
            return dict(data)
        if isinstance(data, (bytes, bytearray)):
            data = bytes(data).decode("utf-8")
        if isinstance(data, str):
            decoded = json.loads(data)
            if not isinstance(decoded, dict):
                raise ValueError("Transaction payload must be a JSON object")
            return decoded
        raise TypeError(f"Unsupported transaction payload type: {type(data)!r}")

    def to_dict(self) -> dict:
        return dict(self.payload)

    # ------------------------------------------------------------------ #
    # Accessors used by the node for validation
    # ------------------------------------------------------------------ #

    @property
    def sender(self):
        return self.payload.get("sender")

    @property
    def sender_address(self):
        sender = self.sender
        if not isinstance(sender, str) or not sender:
            return None
        return try_address_from_public_key(sender)

    @property
    def nonce(self):
        return self.payload.get("nonce")

    @property
    def tx_id(self) -> str:
        """Replay-protection id: hash of the signed part of the payload."""
        return hashlib.sha256(self.serialize_message(self.payload)).hexdigest()

    def value_units(self) -> int:
        if "value" not in self.payload:
            return 0
        return san_to_units(self.payload["value"])

    # ------------------------------------------------------------------ #
    # Canonical serialization
    # ------------------------------------------------------------------ #

    @staticmethod
    def serialize_message(message: dict) -> bytes:
        """
        Serializes the message in canonical JSON format.

        ``signature`` and ``fee`` are protocol metadata: they are excluded so
        that signing and verification always reconstruct the exact same bytes.
        """
        message_to_sign = {
            key: value
            for key, value in message.items()
            if key not in Transaction.META_FIELDS
        }
        return canonical.dumps_bytes(message_to_sign)

    @staticmethod
    def _fee_basis(payload: dict) -> bytes:
        """Canonical bytes the protocol fee is metered on.

        Includes the signature (the real serialized size of a transaction) but
        excludes the protocol-set ``fee`` field, so every node can recompute the
        exact same fee.
        """
        basis = {key: value for key, value in payload.items() if key != "fee"}
        return canonical.dumps_bytes(basis)

    @staticmethod
    def expected_fee(payload: dict, fee_rate: int) -> int:
        """Deterministic protocol fee: size fee plus maximum gas cost.

        The full gas escrow (``gas_limit * gas_price``) is charged up front and
        the unused part is refunded after execution, so every node can verify
        the fee without running the code first.
        """
        size_fee = len(Transaction._fee_basis(payload)) * int(fee_rate)
        try:
            gas_limit = Transaction._int_field(payload, "gas_limit")
            gas_price = Transaction._int_field(payload, "gas_price")
        except ValueError:
            gas_limit = 0
            gas_price = 0
        return size_fee + gas_limit * gas_price

    @staticmethod
    def sign_payload(payload: dict, private_key: bytes) -> str:
        """Sign a transaction payload and return the hex signature.

        Convenience helper so clients and tests use exactly the same canonical
        representation as the verifier.
        """
        return crypto.sign(Transaction.serialize_message(payload), private_key).hex()

    # ------------------------------------------------------------------ #
    # Verification
    # ------------------------------------------------------------------ #

    @staticmethod
    def verify_transaction(tx_bytes) -> bool:
        """
        Verifies a transaction signature with the post-quantum backend.

        The transaction data must contain "sender" (the public key in hex) and
        "signature" (hex). Returns True only when the signature is valid;
        malformed or unsigned data returns False instead of raising, so a single
        bad transaction can never crash a block validator.
        """
        try:
            if isinstance(tx_bytes, (bytes, bytearray)):
                tx = json.loads(bytes(tx_bytes).decode("utf-8"))
            elif isinstance(tx_bytes, str):
                tx = json.loads(tx_bytes)
            elif isinstance(tx_bytes, dict):
                tx = tx_bytes
            else:
                return False
        except Exception as exc:  # noqa: BLE001 - never trust network input
            logger.warning("Transaction decoding failed: %s", exc)
            return False

        if not isinstance(tx, dict) or "sender" not in tx or "signature" not in tx:
            logger.warning("Transaction is missing 'sender' or 'signature'")
            return False

        try:
            public_key = bytes.fromhex(tx["sender"])
            signature = bytes.fromhex(tx["signature"])
        except Exception as exc:  # noqa: BLE001
            logger.warning("Public key/signature conversion failed: %s", exc)
            return False

        message_bytes = Transaction.serialize_message(tx)
        return crypto.verify(message_bytes, signature, public_key)

File: internal/ledger/transaction.go
package ledger

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math/big"

	"github.com/alibertay/san_network/internal/canonical"
	"github.com/alibertay/san_network/internal/crypto"
)

// Transaction mirrors blockchain/Transaction.py. The payload is a Python-like
// dynamic JSON object (map[string]any) so canonical serialization is
// byte-for-byte identical to the reference implementation.

// Fields that are protocol metadata and are never part of a user signature.
var TransactionMetaFields = []string{"signature", "fee"}

// Payload keys that trigger code execution (and therefore need gas).
var executionFields = []string{"bytecode", "contract_code"}

var validatorCommands = map[string]bool{
	"deposit": true, "undelegate": true, "withdraw": true, "evidence": true,
}

var governanceCommands = map[string]bool{"set_param": true}

// Transaction is a normalized, fee-stamped payload.
type Transaction struct {
	Payload map[string]any
	Fee     int64
	Data    []byte
}

// NewTransaction normalizes, validates and fee-stamps a payload.
func NewTransaction(data any, feeRate *int64) (*Transaction, error) {
	payload, err := NormalizePayload(data)
	if err != nil {
		return nil, err
	}
	delete(payload, "fee")

	if !truthy(ChainIDOf(payload)) {
		return nil, fmt.Errorf("Transaction payload must include a non-empty 'chain_id'")
	}
	if message := ValidateGasFields(payload); message != "" {
		return nil, fmt.Errorf("%s", message)
	}
	if message := ValidateValidatorCommand(payload); message != "" {
		return nil, fmt.Errorf("%s", message)
	}
	if message := ValidateGovernanceCommand(payload); message != "" {
		return nil, fmt.Errorf("%s", message)
	}
	if message := ValidateContractCode(payload); message != "" {
		return nil, fmt.Errorf("%s", message)
	}

	rate := MinFeePerByte
	if feeRate != nil {
		rate = *feeRate
	}
	fee := ExpectedFee(payload, rate)
	payload["fee"] = fee
	encoded, err := canonical.Marshal(payload)
	if err != nil {
		return nil, err
	}
	return &Transaction{Payload: payload, Fee: fee, Data: encoded}, nil
}

// NormalizePayload accepts a map, a JSON string or JSON bytes.
func NormalizePayload(data any) (map[string]any, error) {
	switch value := data.(type) {
	case map[string]any:
		copy := make(map[string]any, len(value))
		for key, item := range value {
			copy[key] = item
		}
		return copy, nil
	case []byte:
		return decodePayload(value)
	case string:
		return decodePayload([]byte(value))
	default:
		return nil, fmt.Errorf("Unsupported transaction payload type: %T", data)
	}
}

func decodePayload(raw []byte) (map[string]any, error) {
	decoded, err := canonical.Decode(raw)
	if err != nil {
		return nil, err
	}
	object, ok := decoded.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("Transaction payload must be a JSON object")
	}
	return object, nil
}

// ChainIDOf returns the chain binding field.
func ChainIDOf(payload map[string]any) any {
	return payload["chain_id"]
}

// HasExecution reports whether the payload triggers code execution.
func HasExecution(payload map[string]any) bool {
	for _, field := range executionFields {
		if _, ok := payload[field]; ok {
			return true
		}
	}
	return false
}

func intField(payload map[string]any, name string) (int64, error) {
	value, ok := payload[name]
	if !ok || value == nil {
		return 0, nil
	}
	switch number := value.(type) {
	case bool:
		return 0, fmt.Errorf("'%s' must be a non-negative integer", name)
	case int:
		if number < 0 {
			return 0, fmt.Errorf("'%s' must be a non-negative integer", name)
		}
		return int64(number), nil
	case int64:
		if number < 0 {
			return 0, fmt.Errorf("'%s' must be a non-negative integer", name)
		}
		return number, nil
	case int32:
		if number < 0 {
			return 0, fmt.Errorf("'%s' must be a non-negative integer", name)
		}
		return int64(number), nil
	case float64:
		if number < 0 {
			return 0, fmt.Errorf("'%s' must be a non-negative integer", name)
		}
		return int64(number), nil
	case *big.Int:
		if number == nil || number.Sign() < 0 || !number.IsInt64() {
			return 0, fmt.Errorf("'%s' must be a non-negative integer", name)
		}
		return number.Int64(), nil
	default:
		return 0, fmt.Errorf("'%s' must be a non-negative integer", name)
	}
}

// GasLimitOf returns the payload gas limit.
func GasLimitOf(payload map[string]any) int64 {
	value, _ := intField(payload, "gas_limit")
	return value
}

// GasPriceOf returns the payload gas price.
func GasPriceOf(payload map[string]any) int64 {
	value, _ := intField(payload, "gas_price")
	return value
}

// ValidateGasFields returns an error message or "".
func ValidateGasFields(payload map[string]any) string {
	gasLimit, err := intField(payload, "gas_limit")
	if err != nil {
		return err.Error()
	}
	gasPrice, err := intField(payload, "gas_price")
	if err != nil {
		return err.Error()
	}
	if HasExecution(payload) {
		if gasLimit < 1 {
			return "execution transactions require 'gas_limit' >= 1"
		}
		if gasPrice < MinGasPrice {
			return fmt.Sprintf("'gas_price' must be at least %d", MinGasPrice)
		}
		return ""
	}
	if gasLimit != 0 || gasPrice != 0 {
		return "transactions without execution must not set gas fields"
	}
	return ""
}

// ValidateValidatorCommand applies the structural rules for staking/slashing
// system transactions.
func ValidateValidatorCommand(payload map[string]any) string {
	command, ok := payload["validator"].(map[string]any)
	if !ok {
		if payload["validator"] == nil {
			return ""
		}
		return "'validator' must be an object"
	}
	name, _ := command["command"].(string)
	if !validatorCommands[name] {
		return fmt.Sprintf("unknown validator command: %v", command["command"])
	}
	if HasExecution(payload) {
		return "validator commands must not contain execution payloads"
	}
	if _, ok := payload["value"]; ok {
		return "validator commands must not be transfers"
	}
	if _, ok := payload["receiver"]; ok {
		return "validator commands must not be transfers"
	}

	if name == "deposit" {
		amount, ok := command["amount"].(int64)
		if !ok || amount <= 0 {
			return "validator deposit requires a positive integer 'amount'"
		}
	}
	if name == "evidence" {
		for _, field := range []string{"vote_a", "vote_b"} {
			vote, ok := command[field].(map[string]any)
			if !ok {
				return fmt.Sprintf("evidence requires a '%s' vote object", field)
			}
			if !truthy(vote["public_key"]) || !truthy(vote["signature"]) {
				return fmt.Sprintf("evidence '%s' must be a signed vote", field)
			}
		}
	}
	return ""
}

// ValidateGovernanceCommand applies the structural rules for on-chain
// parameter changes.
func ValidateGovernanceCommand(payload map[string]any) string {
	command, ok := payload["governance"].(map[string]any)
	if !ok {
		if payload["governance"] == nil {
			return ""
		}
		return "'governance' must be an object"
	}
	name, _ := command["command"].(string)
	if !governanceCommands[name] {
		return fmt.Sprintf("unknown governance command: %v", command["command"])
	}
	if HasExecution(payload) {
		return "governance commands must not contain execution payloads"
	}
	if _, ok := payload["value"]; ok {
		return "governance commands must not be transfers"
	}
	if _, ok := payload["receiver"]; ok {
		return "governance commands must not be transfers"
	}
	if payload["validator"] != nil {
		return "governance and validator commands cannot be combined"
	}
	if !truthy(command["name"]) {
		return "governance set_param requires a 'name'"
	}
	if _, ok := command["value"].(int64); !ok {
		return "governance set_param requires an integer 'value'"
	}
	approvals, ok := command["approvals"].([]any)
	if !ok || len(approvals) == 0 {
		return "governance set_param requires at least one approval"
	}
	for _, raw := range approvals {
		approval, ok := raw.(map[string]any)
		if !ok {
			return "governance approvals must be objects"
		}
		if !truthy(approval["public_key"]) || !truthy(approval["signature"]) {
			return "governance approvals must be signed by validators"
		}
	}
	return ""
}

// ValidateContractCode applies the structural rules for deploy/run payloads.
func ValidateContractCode(payload map[string]any) string {
	code, ok := payload["contract_code"].(map[string]any)
	if !ok {
		if payload["contract_code"] == nil {
			return ""
		}
		return "'contract_code' must be an object"
	}
	command, _ := code["command"].(string)
	if command != "deploy" && command != "run" {
		return fmt.Sprintf("unknown contract command: %v", code["command"])
	}
	contractID, ok := code["contract_id"].(string)
	if !ok || contractID == "" {
		return "contract requires a non-empty string 'contract_id'"
	}
	if command == "deploy" {
		_, hasPena := code["pena_code"].(string)
		_, hasBytecode := code["bytecode"].([]any)
		if !hasPena && !hasBytecode {
			return "deploy requires 'pena_code' or 'bytecode'"
		}
		return ""
	}
	functionName, ok := code["function_name"].(string)
	if !ok || functionName == "" {
		return "run requires a non-empty 'function_name'"
	}
	if params, ok := code["params"]; ok && params != nil {
		if _, ok := params.([]any); !ok {
			return "'params' must be a list"
		}
	}
	return ""
}

// SerializeMessage serializes the signed part of a payload (signature and fee
// are excluded).
func SerializeMessage(message map[string]any) ([]byte, error) {
	toSign := make(map[string]any, len(message))
	for key, value := range message {
		if key == "signature" || key == "fee" {
			continue
		}
		toSign[key] = value
	}
	return canonical.Marshal(toSign)
}

func feeBasis(payload map[string]any) ([]byte, error) {
	basis := make(map[string]any, len(payload))
	for key, value := range payload {
		if key == "fee" {
			continue
		}
		basis[key] = value
	}
	return canonical.Marshal(basis)
}

// ExpectedFee is the deterministic protocol fee: size fee plus maximum gas
// cost.
func ExpectedFee(payload map[string]any, feeRate int64) int64 {
	basis, err := feeBasis(payload)
	sizeFee := int64(0)
	if err == nil {
		sizeFee = int64(len(basis)) * feeRate
	}
	gasLimit, errLimit := intField(payload, "gas_limit")
	if errLimit != nil {
		gasLimit = 0
	}
	gasPrice, errPrice := intField(payload, "gas_price")
	if errPrice != nil {
		gasPrice = 0
	}
	return sizeFee + gasLimit*gasPrice
}

// TxID is the replay-protection id: hash of the signed part of the payload.
func TxID(payload map[string]any) string {
	message, err := SerializeMessage(payload)
	if err != nil {
		return ""
	}
	digest := sha256.Sum256(message)
	return hex.EncodeToString(digest[:])
}

// SignPayload signs a payload and returns the hex signature.
func SignPayload(payload map[string]any, privateKey []byte) (string, error) {
	message, err := SerializeMessage(payload)
	if err != nil {
		return "", err
	}
	signature, err := crypto.Sign(message, privateKey)
	if err != nil {
		return "", err
	}
	return hex.EncodeToString(signature), nil
}

// VerifyTransaction verifies a transaction signature. Malformed or unsigned
// data returns false instead of panicking.
func VerifyTransaction(data any) bool {
	var tx map[string]any
	switch value := data.(type) {
	case []byte:
		decoded, err := canonical.Decode(value)
		if err != nil {
			return false
		}
		object, ok := decoded.(map[string]any)
		if !ok {
			return false
		}
		tx = object
	case string:
		decoded, err := canonical.Decode([]byte(value))
		if err != nil {
			return false
		}
		object, ok := decoded.(map[string]any)
		if !ok {
			return false
		}
		tx = object
	case map[string]any:
		tx = value
	default:
		return false
	}

	sender, senderOK := tx["sender"].(string)
	signatureHex, signatureOK := tx["signature"].(string)
	if !senderOK || !signatureOK {
		return false
	}
	publicKey, err := hex.DecodeString(sender)
	if err != nil {
		return false
	}
	signature, err := hex.DecodeString(signatureHex)
	if err != nil {
		return false
	}
	message, err := SerializeMessage(tx)
	if err != nil {
		return false
	}
	return crypto.Verify(message, signature, publicKey)
}

func truthy(value any) bool {
	switch typed := value.(type) {
	case nil:
		return false
	case string:
		return typed != ""
	case bool:
		return typed
	case int64:
		return typed != 0
	case int:
		return typed != 0
	case float64:
		return typed != 0
	case map[string]any:
		return len(typed) > 0
	case []any:
		return len(typed) > 0
	default:
		return true
	}
}

// EncodePayload is the canonical JSON string of a payload (test helper).
func EncodePayload(payload map[string]any) (string, error) {
	data, err := canonical.Marshal(payload)
	if err != nil {
		return "", err
	}
	return string(data), nil
}
